### ISSUE TITLE
Sorting pages with a parent_page_types restriction raises a MovePagePermissionError

### ISSUE THREAD

**Issue by thijskramer** [wagtail/wagtail#8004]
### Issue Summary

If you have a list of pages that are restricted to a certain parent page type and you toggle the sorting via "sort menu order", sorting raises a MovePagePermissionError.

### Steps to Reproduce

1. Start a new project with `wagtail start myproject`
2. Edit home/models.py, and add another page type: 
```
class AnotherPage(Page):
    parent_page_types = ["home.HomePage"]
```
3. run `python manage.py makemigrations` and `python manage.py migrate`
4. Go to the Wagtail Admin, and add a few child pages under the `Home` page
5. While viewing the contents of the `Home` page, toggle the page sorting option via More -> Sort Menu Order:
<img width="544" alt="image" src="https://user-images.githubusercontent.com/1274867/154341965-42325889-72f1-43c0-9a58-549b12a8422c.png">
6. Try to reorder the pages. The order is not being saved. My local development server outputs the following: 

```
raise MovePagePermissionError(
wagtail.core.actions.move_page.MovePagePermissionError: You do not have permission to move the page to the target specified.
```

If I remove the parent_page_types restriction from the model definition, everything works as expected.

* I have confirmed that this issue can be reproduced as described on a fresh Wagtail project: (**yes** / ~no~)

### Technical details

* Python version: 3.9.7
* Django version: 3.2.12 and 4.0.2
* Wagtail version: 2.16.1
* Browser version: Firefox 96.0.3

**Comment by robmoorman** [comment]
Stumpled on kind of the same issue here as well, have a site with a lot of translations and (parent) page types. In many cases I can't sort orders as well.

Needs some deaper investigation, will report it here.

**Comment by robmoorman** [comment]
Update: `parent.specific_class` is wrong (it's the same as the `cls` in this case. But it's definitely not the case.

The `parent` is wrong is this case. It's not the parent but the `target` page (left/right pos)...

**Comment by thijskramer** [comment]
But the fact that you are sorting proves that the moving page _can_ exist under its parent, otherwise it wouldn't be possible to create it in the first place. So why is that method being called when sorting?

**Comment by robmoorman** [comment]
Looks like the `PagePermissionTester` is not setting the parent/target right (maybe with the `pos` in mind)... still don't have a clue why this is for certain page trees and not for all. Can't see any consistency over multiple sites (the same structure DOES work for another site here... very odd).

**Comment by robmoorman** [comment]
Found that when `position_page` is `None` it positions the page. But when a position/page is found, it does not:

In `def set_page_position`:

```python
# Find page that's already in this position
position_page = None
if position is not None:
    try:
        position_page = parent_page.get_children()[int(position)]
    except IndexError:
        pass  # No page in this position
```

**Comment by robmoorman** [comment]
Not yet @thijskramer lacking time here this week. If you need something plz contact me!

**Comment by thijskramer** [comment]
@robmoorman could you please review my Draft PR? The implementation is done (I think) but the tests aren't finished yet, I might finish them later today if I find the time!

**Comment by robmoorman** [comment]
Quick review @thijskramer , important note here is that a setup with:

```
root_page
    SimplePage
        SimplePage
        SimplePage
```

does work currently.

We should cover that this page type hierarchy keeps working as well in the new fix (not only test the new business/child setup).

**Comment by thijskramer** [comment]
Is there anything I have to do that prevents this PR from being merged?

**Comment by cnk** [comment]
Merged fix onto main and 2.16.x
